onnxruntime/test/providers/provider_test_utils.cc

```diff
@@ -194,7 +207,12 @@ void Check<BFloat16>(const OpTester::Data& expected_data, const Tensor& output_t
       EXPECT_EQ(f_expected[i], f_output[i]);
     else {
       // the default for existing tests
-      EXPECT_NEAR(f_expected[i], f_output[i], threshold) << "provider_type: " << provider_type;
+      const float max_value = fmax(fabs(f_expected[i]), fabs(f_output[i]));
+      if (max_value != 0)  // max_value = 0 means output and expected are 0s.
+      {
+        const float rel_error = fabs(f_expected[i] - f_output[i]) / max_value;
+        EXPECT_NEAR(0, rel_error, threshold) << "provider_type: " << provider_type;
+      }
     }
```

Comment:
A comment explaining why we take these extra steps instead of the old approach of just calling EXPECT_NEAR would be helpful.

Reply:
For floating-point numbers, we need to check the relative errors. Absolute values are not reliable. Say, if I implement an `Add` operator using `Sub` logic (so my `Add` is actually `Sub`) and use inputs `0.000000000000000001` and `0.000000000000000009` in my test, I believe it will still contribute an OK test and the stupid me may think my implementation is correct. If we check the result's relative error, we can avoid that.